### PR TITLE
Introduce checksum for uncompressed data in the header

### PR DESCRIPTION
**Pull request by sbabic**
Hi Jonathan,
please consider this as RFC and after just testing that the uncompressed checksum is part of the index and it is reported back by zck_read_header, and after running zck / unzck on quite big files.
This set implements the new format as specified in the uncompressed_source branch. The checksum for uncompressed data is stored then in the index and the hash is computed together with the one for the compressed chunk.
In the series there are also:

Drop warning when OLD_ZSTD is set
----------------------------------------------------- 
It is a nitpick, this just drop a warning when OLD_ZSTD is set

Drop assert in zmalloc and raise runtime error 
--------------------------------------------------------------

This is an issue using the library on embedded system. An allocation on the heap could fail on small system, but the process is not allowed to fail and it should manage this. So I replace the assert with error handling in case of zmalloc(), same should be done for zrealloc(), but I want to know first your opinion on thisand if can be mainlined.

Thanks,
Stefano

### CODE REVIEW

**Comment by sbabic** [comment]
Note: checksum for uncompressed data is activated in zck with -u parameter: 
zck -vvvvv -u terminal-s

**Comment by jdieter** [comment]
Thanks so much for the PR!

I'm on holiday right now, and have really crappy internet and no laptop, so
I'll have to wait until next week to review this.

File: src/zck_read_header.c
Comment (jdieter) [inline src/zck_read_header.c:183]: Is this line needed?

File: src/zck_read_header.c
Comment (jdieter) [inline src/zck_read_header.c:184]: It would be nice to have a field header "Uncompressed Chunk Checksum" when the flag is set.

**Review by jdieter — CHANGES_REQUESTED** [review]
This looks really good.  I'm happy enough to drop the assert on zmalloc, and, instead return a run-time error.  As you've noticed, it will take a lot of grunt work to get it in there, but, if you're happy to write the code, I'm happy to maintain it.

I requested a few minor changes, but I'd also like to see some test cases exercising the code.

Thanks again for this contribution!  I look forward into getting this feature into zchunk.

File: src/lib/comp/comp.c
Comment (sbabic) [reply src/lib/comp/comp.c]: Fixed, thanks !

File: src/lib/index/index_create.c
Comment (sbabic) [reply src/lib/index/index_create.c]: I have to remember that you are using spaces as code styling. I have to set vim with noexpandtab before working on zck.

Fixed.

File: src/lib/index/index_read.c
Comment (sbabic) [reply src/lib/index/index_read.c]: Done

File: src/zck_read_header.c
Comment (sbabic) [reply src/zck_read_header.c]: No, sorry, a debug output I left. Drop it.

File: src/zck_read_header.c
Comment (sbabic) [reply src/zck_read_header.c]: Done.

**Comment by sbabic** [comment]
I have not added any further test because I have not found in /test some use cases with usage of zck_write(). Code is really tested using the built zck, that is running "zck -u" and then zck_read_header. It should be enough to run it at shell script instead of C code.
By the way, I am thinking about to go on with the following topics:
- check occurrencies of realloc and calloc and convert them to runtime errors
- avoid to allocate the whole file in RAM. Currently, the file to be converted is entirely loaded in RAM. This is not possible with large file, for example if a filesystem as basis is used. I want to convert this and let zck_write() gets data reading on demand from file instead of loading into RAM. (zck_write() works already using a stream-like, without seeks).
- I added uthash for uncompression digest in this series - anyway, I do not want to add a lot of things now until the first PR get be merged.

Thanks for your help and maintaining this project !

**Comment by jdieter** [comment]
Regarding your overall comments, yes, the tests are mainly integration tests that just run the commands.  If you would like me to write the tests, I can do so.

I would be happy to have to change realloc and calloc to run-time errors.

I would also love to have the code fixed so we're not reading the full file into RAM when creating the zchunk file.  It should be a fairly simple fix, but it's complicated by the fact that we allow the user to provide a string to split the file on, rather than use auto-chunking, which means we have to cache at least the last `string_length` of each read to make sure we don't miss a potential match on the read border.

Once you've addressed the last couple of issues, I'll go ahead and merge this PR, but I won't release a new version of zchunk until we have some tests and `unzck -u <uncompressed source>` works.

**Review by jdieter — CHANGES_REQUESTED** [review]

File: src/lib/comp/comp.c
Comment (sbabic) [reply src/lib/comp/comp.c:598]: I will retest and check again

File: src/zck_read_header.c
Comment (sbabic) [reply src/zck_read_header.c:185]: Sure, I' ll do for next version.

**Comment by sbabic** [comment]
> Regarding your overall comments, yes, the tests are mainly integration tests that just run the commands. If you would like me to write the tests, I can do so.

I am not sure what you really want, so yes, if you can do it this time, I will probably take over next time.
> 
> I would be happy to have to change realloc and calloc to run-time errors.

I prepare this and I will post after this PR will be merged.

> 
> I would also love to have the code fixed so we're not reading the full file into RAM when creating the zchunk file. It should be a fairly simple fix, but it's complicated by the fact that we allow the user to provide a string to split the file on, rather than use auto-chunking, which means we have to cache at least the last `string_length` of each read to make sure we don't miss a potential match on the read border.

In fact, my proposal is to add it just in case of automatic chunks, and raises an error if this feature is activated in other cases, letting current behaviour unchanged (feature must be turned on).

> 
> Once you've addressed the last couple of issues, I'll go ahead and merge this PR,

Fine, thanks !

> but I won't release a new version of zchunk until we have some tests and `unzck -u <uncompressed source>` works.

This is fine with me, thanks.

File: src/lib/comp/comp.c
Comment (sbabic) [reply src/lib/comp/comp.c:598]: For reason I have not yet understood, the check is depending on the compression. Just disabling thwe check as in patch I get the results I am expecting. What I am doing:
- I run zck with -u option on a file (650MB)
- I have written a piece of code to call the library to generate the header from the same file, but without compression. That means I call zck_set_ioption(zck, ZCK_COMP_TYPE, ZCK_COMP_NONE) and zck_set_ioption(zck, ZCK_UNCOMP_HEADER, 1) before zck_write()
To be able to decide later which chunks should be installed, in both cases  I should get the same number of chunks (because I am using the same source file for testing) and chunks should have the same digest for uncompressed data.
This is what I get with my patch. However, if not applied, even if both are running comp_init() (from zck_write) and setting the same values for min and max in autochunk mode, zck (compressing with zst) reports a lot of "small chunks", and the two results are quite different.

I guessed I could have the same effect by zck_set_ioption(zck, ZCKCHUNK_MIN, 1). However, this runs in a chicken egg issue. It should be possible to set this before starting, that means before calling zck_write. This is not true because default value are set later by zck_write (in comp_init), and if I call it before zck_write I get:

comp_ioption: Minimum chunk size must be <= maximum chunk size
Minimum chunk size must be <= maximum chunk size

This is clear from code, but IMO the check should be done later (after zck_write is called) and not when setting during zck_set_option(). For example, I could require to set ZCKCHUNK_MIN but not ZCKCHUNK_MAX or viceversa.

**Comment by sbabic** [comment]
New version with : 
- fixes as requested
- added commit to replace assert in case of OOM (for zrealloc and calloc in buzhash_update)
Commit for disablich minimum check size is still there - even if the check should be on decompressed data (and then the chunk size should be independent from the chosen compressor), results are different with zstd or nocomp. Just disabling at all I get the same results.

**Review by jdieter — APPROVED** [review]
This looks good, so I'm going to merge it, but there are some changes I'd like to see before we release zchunk-1.2:
* Figure out why we need ZCK_NO_MIN_CHUNKSIZE and see if we can revert that
* Add support for unzck -u
* Add tests
* Stop reading full file into RAM when creating the zchunk file

**Comment by sbabic** [comment]
> This looks good, so I'm going to merge it, but there are some changes I'd like to see before we 4release zchunk-1.2:
> 

Thanks !

>     * Figure out why we need ZCK_NO_MIN_CHUNKSIZE and see if we can revert that
> 
>     * Add support for unzck -u
> 
>     * Add tests
> 
>     * Stop reading full file into RAM when creating the zchunk file

Confirm I will work on last topic, at least in the case when autochunk is used.

**Pull request merged**
